src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py

```diff
@@ -215,29 +244,70 @@ def visit_StencilClosure(self, node: itir.StencilClosure, **kwargs: Any) -> Sten
         )
 
         return StencilExecution(
-            stencil=self.visit(node.stencil),
+            stencil=self.visit(node.stencil, **kwargs),
             output=self._visit_output_argument(node.output),
             inputs=self.visit(node.inputs),
             backend=backend,
         )
 
+    @staticmethod
+    def _get_gridtype(closures: list[itir.StencilClosure]) -> common.GridType:
+        def _extract_grid_type(domain_arg: itir.Expr) -> common.GridType:
+            assert isinstance(domain_arg, itir.FunCall)
+            if domain_arg.fun == itir.SymRef(id="cartesian_domain"):
+                return common.GridType.CARTESIAN
+            else:
+                assert domain_arg.fun == itir.SymRef(id="unstructured_domain")
+                return common.GridType.UNSTRUCTURED
+
+        first = _extract_grid_type(closures[0].domain)
+        assert all(first == _extract_grid_type(c.domain) for c in closures)
+        return first
+
+    def _collect_dimensions_from_domain(self, closures: list[itir.StencilClosure]) -> None:
+        for c in closures:
+            if self.grid_type == common.GridType.CARTESIAN:
+                if isinstance(c.domain, itir.FunCall) and c.domain.fun == itir.SymRef(
+                    id="cartesian_domain"
+                ):
+                    for nr in c.domain.args:
+                        assert isinstance(nr, itir.FunCall) and nr.fun == itir.SymRef(
+                            id="named_range"
+                        )
+                        dim_name = nr.args[0].value
+                        self.offset_definitions[dim_name] = OffsetDefinition(name=Sym(id=dim_name))
+            else:
+                if isinstance(c.domain, itir.FunCall) and c.domain.fun == itir.SymRef(
+                    id="unstructured_domain"
+                ):
+                    if len(c.domain.args) > 0:
+                        horizontal_name = c.domain.args[0].args[0].value
+                        self.offset_definitions[horizontal_name] = OffsetDefinition(
+                            name=Sym(id=horizontal_name), alias=self._horizontal_dimension
+                        )
+                    if len(c.domain.args) > 1:
+                        vertical_name = c.domain.args[1].args[0].value
+                        self.offset_definitions[vertical_name] = OffsetDefinition(
+                            name=Sym(id=vertical_name), alias=self._vertical_dimension
+                        )
+                    if len(c.domain.args) > 2:
```

Comment:
I kept an exception as it can be a user error, but changed to ValueError. We should do a refactoring of the whole library to use (custom) exceptions consistently.